README.md
[![Build Status](https://travis-ci.org/pamit/persian_numbers.svg?branch=master)](https://travis-ci.org/pamit/persian_numbers)
[![Coverage Status](https://coveralls.io/repos/github/pamit/persian_numbers/badge.svg?branch=master)](https://coveralls.io/github/pamit/persian_numbers?branch=master)
[![Gem Version](https://badge.fury.io/rb/persian_numbers.svg)](https://badge.fury.io/rb/persian_numbers)

# PersianNumbers

This gem serves as a set of utility functions for Persian numbers, such as conversion of numeric inputs to their equivalent values in Persian words.


## Installation

Add this line to your application's Gemfile:

```ruby
gem 'persian_numbers'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install persian_numbers


## Usage

To convert a number to it equivalent Persian words:

```ruby
PersianNumbers.translate_to_words(35)
# سی و پنج

PersianNumbers.translate_to_words(12345)
# دوازده هزار و سیصد و چهل و پنج
```

You can also run the following command to see a demo:

```ruby
ruby demo/generate.rb
```

## CHANGELOG

[https://github.com/pamit/persian_numbers/CHANGELOG.md](https://github.com/pamit/persian_numbers/CHANGELOG.md)


## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).


## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/pamit/persian_numbers. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/pamit/persian_numbers/blob/master/CODE_OF_CONDUCT.md).


## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


## Code of Conduct

Everyone interacting in the PersianNumbers project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/pamit/persian_numbers/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
### 1.0.0 - 2020-06-10

* features
  * Add `translate_to_words` method to convert a numeric input to its equivalent Persian words.

spec/spec_helper.rb
require 'coveralls'
Coveralls.wear!

require 'bundler/setup'
require 'persian_numbers'

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/persian_numbers_spec.rb
RSpec.describe PersianNumbers do
  describe '.translate_to_words' do
    context 'with invalid input' do
      it {
        expect { described_class.translate_to_words('INVALID') }
          .to raise_error(Conversion::InvalidNumberError)
      }
    end

    context 'with 1 digit' do
      it { expect(described_class.translate_to_words(9)).to eq('نه') }
      it { expect(described_class.translate_to_words('9')).to eq('نه') }
      it { expect(described_class.translate_to_words('۹')).to eq('نه') }

      it { expect(described_class.translate_to_words('09')).to eq('نه') }
    end

    context 'with 2 digits' do
      it { expect(described_class.translate_to_words(35)).to eq('سی و پنج') }
      it { expect(described_class.translate_to_words('35')).to eq('سی و پنج') }
      it { expect(described_class.translate_to_words('۳۵')).to eq('سی و پنج') }

      it { expect(described_class.translate_to_words('0035')).to eq('سی و پنج') }
    end

    context 'with 3 digits' do
      it { expect(described_class.translate_to_words(456)).to eq('چهارصد و پنجاه و شش') }
      it { expect(described_class.translate_to_words('456')).to eq('چهارصد و پنجاه و شش') }
      it { expect(described_class.translate_to_words('۴۵۶')).to eq('چهارصد و پنجاه و شش') }

      it { expect(described_class.translate_to_words(400)).to eq('چهارصد') }
      it { expect(described_class.translate_to_words(410)).to eq('چهارصد و ده') }
      it { expect(described_class.translate_to_words(417)).to eq('چهارصد و هفده') }
    end

    context 'with 4 digits' do
      it { expect(described_class.translate_to_words(7812)).to eq('هفت هزار و هشتصد و دوازده') }
      it { expect(described_class.translate_to_words('7812')).to eq('هفت هزار و هشتصد و دوازده') }
      it { expect(described_class.translate_to_words('۷۸۱۲')).to eq('هفت هزار و هشتصد و دوازده') }

      it { expect(described_class.translate_to_words(7012)).to eq('هفت هزار و دوازده') }
      it { expect(described_class.translate_to_words(7802)).to eq('هفت هزار و هشتصد و دو') }
      it { expect(described_class.translate_to_words(7000)).to eq('هفت هزار') }
      it { expect(described_class.translate_to_words(7004)).to eq('هفت هزار و چهار') }
      it { expect(described_class.translate_to_words(7031)).to eq('هفت هزار و سی و یک') }
      it { expect(described_class.translate_to_words(7630)).to eq('هفت هزار و ششصد و سی') }
    end

    context 'with 5 digits' do
      it { expect(described_class.translate_to_words(12_345)).to eq('دوازده هزار و سیصد و چهل و پنج') }
      it { expect(described_class.translate_to_words('12345')).to eq('دوازده هزار و سیصد و چهل و پنج') }
      it { expect(described_class.translate_to_words('۱۲۳۴۵')).to eq('دوازده هزار و سیصد و چهل و پنج') }

      it { expect(described_class.translate_to_words(12_045)).to eq('دوازده هزار و چهل و پنج') }
      it { expect(described_class.translate_to_words(22_315)).to eq('بیست و دو هزار و سیصد و پانزده') }
      it { expect(described_class.translate_to_words(40_315)).to eq('چهل هزار و سیصد و پانزده') }
      it { expect(described_class.translate_to_words(41_307)).to eq('چهل و یک هزار و سیصد و هفت') }
      it { expect(described_class.translate_to_words(40_300)).to eq('چهل هزار و سیصد') }
    end

    context 'with 6 digits' do
      it { expect(described_class.translate_to_words(421_035)).to eq('چهارصد و بیست و یک هزار و سی و پنج') }
      it { expect(described_class.translate_to_words('421035')).to eq('چهارصد و بیست و یک هزار و سی و پنج') }
      it { expect(described_class.translate_to_words('۴۲۱۰۳۵')).to eq('چهارصد و بیست و یک هزار و سی و پنج') }

      it { expect(described_class.translate_to_words(420_035)).to eq('چهارصد و بیست هزار و سی و پنج') }
      it { expect(described_class.translate_to_words(420_005)).to eq('چهارصد و بیست هزار و پنج') }
      it { expect(described_class.translate_to_words(420_105)).to eq('چهارصد و بیست هزار و صد و پنج') }
      it { expect(described_class.translate_to_words(420_115)).to eq('چهارصد و بیست هزار و صد و پانزده') }
      it { expect(described_class.translate_to_words(420_015)).to eq('چهارصد و بیست هزار و پانزده') }
      it { expect(described_class.translate_to_words(400_005)).to eq('چهارصد هزار و پنج') }
    end

    context 'with 7 digits' do
      it { expect(described_class.translate_to_words(5_619_231)).to eq('پنج میلیون و ششصد و نوزده هزار و دویست و سی و یک') }
      it { expect(described_class.translate_to_words('5619231')).to eq('پنج میلیون و ششصد و نوزده هزار و دویست و سی و یک') }
      it { expect(described_class.translate_to_words(5_000_001)).to eq('پنج میلیون و یک') }
      it { expect(described_class.translate_to_words(5_000_011)).to eq('پنج میلیون و یازده') }
      it { expect(described_class.translate_to_words(5_040_011)).to eq('پنج میلیون و چهل هزار و یازده') }
      it { expect(described_class.translate_to_words(5_040_712)).to eq('پنج میلیون و چهل هزار و هفتصد و دوازده') }
      it { expect(described_class.translate_to_words(5_600_712)).to eq('پنج میلیون و ششصد هزار و هفتصد و دوازده') }
      it { expect(described_class.translate_to_words(5_605_712)).to eq('پنج میلیون و ششصد و پنج هزار و هفتصد و دوازده') }
      it { expect(described_class.translate_to_words(5_609_031)).to eq('پنج میلیون و ششصد و نه هزار و سی و یک') }
      it { expect(described_class.translate_to_words(5_609_013)).to eq('پنج میلیون و ششصد و نه هزار و سیزده') }
      it { expect(described_class.translate_to_words(5_019_201)).to eq('پنج میلیون و نوزده هزار و دویست و یک') }
    end

    context 'with 9 digits' do
      it {
        expect(described_class.translate_to_words(312_897_456))
          .to eq('سیصد و دوازده میلیون و هشتصد و نود و هفت هزار و چهارصد و پنجاه و شش')
      }
      it {
        expect(described_class.translate_to_words('312897456'))
          .to eq('سیصد و دوازده میلیون و هشتصد و نود و هفت هزار و چهارصد و پنجاه و شش')
      }
    end

    context 'with 11 digits' do
      it {
        expect(described_class.translate_to_words(45_312_897_456))
          .to eq('چهل و پنج میلیارد و سیصد و دوازده میلیون و هشتصد و نود و هفت هزار و چهارصد و پنجاه و شش')
      }
      it {
        expect(described_class.translate_to_words('45312897456'))
          .to eq('چهل و پنج میلیارد و سیصد و دوازده میلیون و هشتصد و نود و هفت هزار و چهارصد و پنجاه و شش')
      }
    end

    context 'with leading zero' do
      it { expect(described_class.translate_to_words(0o0456)).not_to eq('چهارصد و پنجاه و شش') }
      it { expect(described_class.translate_to_words('00456')).to eq('چهارصد و پنجاه و شش') }
      it { expect(described_class.translate_to_words('00000456')).to eq('چهارصد و پنجاه و شش') }
      it { expect(described_class.translate_to_words('00000450')).to eq('چهارصد و پنجاه') }
    end

    context 'with comma' do
      it {
        expect(described_class.translate_to_words('45,312,897,456'))
          .to eq('چهل و پنج میلیارد و سیصد و دوازده میلیون و هشتصد و نود و هفت هزار و چهارصد و پنجاه و شش')
      }
    end
  end
end


lib/persian_numbers/constants.rb
PERSIAN_NUMBER_MAP = {
  '۰' => '0',
  '۱' => '1',
  '۲' => '2',
  '۳' => '3',
  '۴' => '4',
  '۵' => '5',
  '۶' => '6',
  '۷' => '7',
  '۸' => '8',
  '۹' => '9',
  '.' => '.'
}.freeze

# Refer to:
# https://fa.wikipedia.org/wiki/%D9%86%D8%A7%D9%85_%D8%A7%D8%B9%D8%AF%D8%A7%D8%AF_%D8%A8%D8%B2%D8%B1%DA%AF
# https://en.wikipedia.org/wiki/Names_of_large_numbers
#
LARGE_NUMBERS_IN_PERSIAN = {
  0 => '',
  3 => 'هزار',
  6 => 'میلیون',
  9 => 'میلیارد',
  12 => 'بیلیون',
  15 => 'بیلیارد',
  18 => 'تریلیون',
  21 => 'تریلیارد',
  24 => 'کوآدریلیون',
  27 => 'کادریلیارد',
  30 => 'کوینتیلیون',
  33 => 'کوانتینیارد',
  36 => 'سکستیلیون',
  39 => 'سکستیلیارد',
  42 => 'سپتیلیون',
  45 => 'سپتیلیارد',
  48 => 'اکتیلیون',
  51 => 'اکتیلیارد',
  54 => 'نانیلیون',
  57 => 'نانیلیارد',
  60 => 'دسیلیون',
  63 => 'دسیلیارد',
  66 => 'آندسیلیون',
  69 => 'آندسیلیارد',
  72 => 'دودسیلیون',
  75 => 'دودسیلیارد',
  78 => 'تریدسیلیون',
  81 => 'تریدسیلیارد',
  84 => 'کوادردسیلیون',
  87 => 'کوادردسیلیارد',
  90 => 'کویندسیلیون',
  93 => 'کویندسیلیارد',
  96 => 'سیدسیلیون',
  99 => 'سیدسیلیارد'
}.freeze

PLACE_VALUE_ONES = {
  1 => 'یک',
  2 => 'دو',
  3 => 'سه',
  4 => 'چهار',
  5 => 'پنج',
  6 => 'شش',
  7 => 'هفت',
  8 => 'هشت',
  9 => 'نه'
}.freeze

PLACE_VALUE_TEENS = {
  1 => 'یازده',
  2 => 'دوازده',
  3 => 'سیزده',
  4 => 'چهارده',
  5 => 'پانزده',
  6 => 'شانزده',
  7 => 'هفده',
  8 => 'هجده',
  9 => 'نوزده'
}.freeze

PLACE_VALUE_TENS = {
  1 => 'ده',
  2 => 'بیست',
  3 => 'سی',
  4 => 'چهل',
  5 => 'پنجاه',
  6 => 'شصت',
  7 => 'هفتاد',
  8 => 'هشتاد',
  9 => 'نود'
}.freeze

PLACE_VALUE_HUNDREDS = {
  1 => 'صد',
  2 => 'دویست',
  3 => 'سیصد',
  4 => 'چهارصد',
  5 => 'پانصد',
  6 => 'ششصد',
  7 => 'هفتصد',
  8 => 'هشتصد',
  9 => 'نهصد'
}.freeze


lib/persian_numbers/conversion.rb
class Conversion
  class InvalidNumberError < StandardError; end

  PLACE_VALUE_SEPARATOR = ' و '

  attr_reader :number

  def initialize(number:)
    @number = format_number(number.to_s)
  end

  def translate_to_words
    number = map_persian_input(@number)
    validate_input(number)
    numbers = format_number_with_separator(number)
    translate(numbers)
  rescue StandardError
    raise InvalidNumberError, "Error in parsing and converting number: #{@number}"
  end

  private

  # Converts a Persian number to English format
  # e.g. map_persian_input('۱۲۳۴۵') => 12345
  #
  def map_persian_input(number)
    number.chars
          .map { |ch| PERSIAN_NUMBER_MAP[ch] || ch }
          .join
  end

  # Safe-parses number and raise exception if not valid
  #
  def validate_input(number)
    Integer(number)
  rescue StandardError
    raise InvalidNumberError, 'The input is not a valid number, should be digits.'
  end

  # Formats number and add separator
  # e.g. format_number_with_separator('12345') => 12,345
  #
  def format_number_with_separator(number, separator = ',')
    number
      .reverse
      .gsub(/(\d{3})(?=\d)/, "\\1#{separator}")
      .reverse
      .split(',')
  end

  # Translates number into it Persian word format
  # e.g. translate(["12", "345"]) => دوازده هزار و سیصد و چهل و پنج
  #
  def translate(numbers)
    converted = []

    numbers.each_with_index do |value, index|
      base_name = LARGE_NUMBERS_IN_PERSIAN[(numbers.size - index - 1) * 3]
      place_value = translate_group(value)
      converted << "#{place_value} #{base_name}".strip if place_value != ''
    end

    converted.delete_if(&:empty?).join(PLACE_VALUE_SEPARATOR)
  end

  # Translates each group of 3 digits
  # e.g. translate_group("345") => سیصد و چهل و پنج
  #
  def translate_group(number)
    return '' if convert_to_integer(number) == 0

    converted = []
    length = number.size
    position = length
    number.chars.each_with_index do |value, _index|
      place_value = translate_place_value(number, value, position)
      converted << place_value
      position -= 1
      position -= 1 if PLACE_VALUE_TEENS.values.include?(place_value)
    end

    converted.compact.join(PLACE_VALUE_SEPARATOR)
  end

  # Translates each place value
  # e.g. translate_place_value("3", 3) => سیصد
  #
  def translate_place_value(number, value, position)
    converted = nil
    int_value = Integer(value)

    if position == 1
      converted = PLACE_VALUE_ONES[int_value] if int_value != 0
    elsif position == 2
      last_two_digits = number.split(//).last(2).join
      if int_value == 1 && (11..19).include?(convert_to_integer(last_two_digits))
        converted = PLACE_VALUE_TEENS[Integer(number[-1])]
      elsif int_value != 0
        converted = PLACE_VALUE_TENS[int_value]
      end
    elsif position == 3
      converted = PLACE_VALUE_HUNDREDS[int_value] if int_value != 0
    end

    converted
  end

  # Converts string to integer
  #
  def convert_to_integer(number)
    formatted = format_number(number)
    formatted = '0' if formatted == ''
    Integer(formatted)
  end

  # Removes unnecessary characters
  #
  def format_number(number)
    number
      .gsub(/,/, '')
      .gsub(/^0*/, '')
  end
end
